service/storage_proxy.cc

```diff
@@ -5154,12 +5173,22 @@ void storage_proxy::init_messaging_service(shared_ptr<migration_manager> mm) {
                 unwrapped = ::compat::unwrap(std::move(pr), *s);
                 auto timeout = t ? *t : db::no_timeout;
                 return p->query_mutations_locally(std::move(s), std::move(cmd), unwrapped, timeout, trace_state_ptr);
-            }).finally([&trace_state_ptr, src_ip] () mutable {
+            }).then_wrapped([this, &trace_state_ptr, src_ip] (future<rpc::tuple<foreign_ptr<lw_shared_ptr<reconcilable_result>>, cache_temperature>> f) mutable {
+                using final_tuple_type = rpc::tuple<foreign_ptr<lw_shared_ptr<reconcilable_result>>, cache_temperature, std::optional<replica::exception_variant>>;
                 tracing::trace(trace_state_ptr, "read_mutation_data handling is done, sending a response to /{}", src_ip);
+                if (!f.failed()) {
+                    auto&& [result, temp] = f.get();
+                    return make_ready_future<final_tuple_type>(std::move(result), std::move(temp), std::nullopt);
+                } else if (features().cluster_supports_typed_errors_in_read_rpc()) {
```

Comment:
Why do you want to convert existing exceptions to the new interface? You previous series allowed to propagate some errors as exceptions and others as values. I think here we can do the same. Rate limiting errors can be propagated as values, but all others stay as they are. Also for rate limiting would it be useful to allow propagating rate limiting requests along with a successful reply?

Reply:
Converting exceptions to values protects against exception storms reducing throughput.

Reply:
Yes, but with previous patches we decided that we do it for exceptions that usually cause storms like timeout exception (hence using `std::expected` or similar. Here for instance we do not solve the problem of timeout exception at all since it is reported on a coordinator by the rpc as an exception anyway. I understand that rate limiting can cause storms and it worth to be propagated as a value (BTW it look like it is converted back to an exception right after been reserved).

Another point is that requesting a rate limit can be done together with returning a valid result.

Reply:
This is for timeout exceptions. Not RPC timeouts, but timeouts detected on the replica while waiting. You're right that we need to change RPC to return std::expected (or something similar) too.

I don't understand the second point.

Reply:
> Yes, but with previous patches we decided that we do it for exceptions that usually cause storms like timeout exception (hence using `std::expected` or similar. Here for instance we do not solve the problem of timeout exception at all since it is reported on a coordinator by the rpc as an exception anyway. I understand that rate limiting can cause storms and it worth to be propagated as a value (BTW it look like it is converted back to an exception right after been reserved).

This PR is only concerned with encoding exceptions and transferring them without losing information about the type. I see "rate limit exceeded" as a kind of an error response, similar to "timeout" or others. I thought it would make sense to introduce a common mechanism for reporting them.

I'm aware that I am converting back to exception, I intend to get rid of this problem in later PRs. This one is just about preserving the types.

> Another point is that requesting a rate limit can be done together with returning a valid result.

Maybe, although currently I plan for the "rate limit exceeded" signal to always mean an error without a result. If we add a possibility to request rate limit while returning the result, we will have to signal the second case differently from the first anyway.

Reply:
> This is for timeout exceptions. Not RPC timeouts, but timeouts detected on the replica while waiting. You're right that we need to change RPC to return std::expected (or something similar) too.

Timeouts detected on a replica side are never reach a coordinator. They are dropped by the client when sending a reply since rpc sees that the execution timed out, so for the timeout converting it to something else here is just a waste of cpu time.

> 
> I don't understand the second point.

Rate limit request does not have to be an error. Piotr seams to understood what I mean.

Reply:
> > Yes, but with previous patches we decided that we do it for exceptions that usually cause storms like timeout exception (hence using `std::expected` or similar. Here for instance we do not solve the problem of timeout exception at all since it is reported on a coordinator by the rpc as an exception anyway. I understand that rate limiting can cause storms and it worth to be propagated as a value (BTW it look like it is converted back to an exception right after been reserved).
> 
> This PR is only concerned with encoding exceptions and transferring them without losing information about the type. I see "rate limit exceeded" as a kind of an error response, similar to "timeout" or others. I thought it would make sense to introduce a common mechanism for reporting them.
> 

And timeout exceptions will never be sent back, so converting them just a waste of time. Some others are rare (like request aborted).